src/service/mod.rs

```diff
@@ -0,0 +1,152 @@
+use rocket::async_trait;
+
+pub mod name;
+pub mod s3;
+
+#[async_trait]
+pub trait KeplerService {
+    type Error;
+    type Stopped;
+
+    async fn start(config: Self::Stopped) -> Result<Self, Self::Error>
+    where
+        Self: Sized;
+    async fn stop(self);
+}
+
+mod vec_cid_bin {
+    use libipld::cid::Cid;
+    use serde::{de::Error as DeError, ser::SerializeSeq, Deserialize, Deserializer, Serializer};
+
+    pub fn serialize<S>(vec: &Vec<Cid>, ser: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut seq = ser.serialize_seq(Some(vec.len()))?;
+        for cid in vec {
+            seq.serialize_element(&cid.to_bytes())?;
+        }
+        seq.end()
+    }
+
+    pub fn deserialize<'de, D>(deser: D) -> Result<Vec<Cid>, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let s: Vec<&[u8]> = Deserialize::deserialize(deser)?;
+        s.iter()
+            .map(|&sc| Cid::read_bytes(sc).map_err(D::Error::custom))
+            .collect()
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::name::*;
+    use super::s3::*;
+    use super::*;
+    use ipfs_embed::{generate_keypair, Config, Event as SwarmEvent, Ipfs};
+    use rocket::futures::StreamExt;
+    use std::collections::BTreeMap;
+
+    fn tracing_try_init() {
+        tracing_subscriber::fmt()
+            .with_env_filter(tracing_subscriber::EnvFilter::from_default_env())
+            .try_init()
+            .ok();
+    }
+
+    async fn create_store(id: &str, path: std::path::PathBuf) -> Result<KNSStore, anyhow::Error> {
+        std::fs::create_dir_all(&path)?;
+        let mut config = Config::new(&path, generate_keypair());
+        config.network.broadcast = None;
+        let ipfs = Ipfs::new(config).await?;
+        ipfs.listen_on("/ip4/0.0.0.0/tcp/0".parse()?)?.next().await;
+        let task_ipfs = ipfs.clone();
+        tokio::spawn(async move {
+            let mut events = task_ipfs.swarm_events();
+            loop {
+                match events.next().await {
+                    Some(SwarmEvent::Discovered(p)) => {
+                        tracing::debug!("dialing peer {}", p);
```

Comment:
I recommend to import `tracing::debug` to make it easier to replace it or things like that.